src/sinks/util/statistic.rs

```diff
@@ -21,60 +22,86 @@ pub struct DistributionStatistic {
-        let mut samples = Vec::new();
-        for sample in source {
-            for _ in 0..sample.rate {
-                samples.push(sample.value);
-            }
-        }
+        let mut bins = source
+            .iter()
+            .filter(|sample| sample.rate > 0)
+            .cloned()
+            .collect::<Vec<_>>();
 
-        if samples.is_empty() {
+        if bins.is_empty() {
             return None;
         }
 
-        if samples.len() == 1 {
-            let val = samples[0];
+        if bins.len() == 1 {
+            let val = bins[0].value;
+            let count = bins[0].rate;
             return Some(Self {
                 min: val,
                 max: val,
                 median: val,
                 avg: val,
-                sum: val,
-                count: 1,
+                sum: val * count as f64,
+                count: count as u64,
                 quantiles: quantiles.iter().map(|&p| (p, val)).collect(),
             });
         }
 
-        samples.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Ordering::Equal));
+        bins.sort_unstable_by(|a, b| a.value.partial_cmp(&b.value).unwrap_or(Ordering::Equal));
+
+        let min = bins.first().unwrap().value;
+        let max = bins.last().unwrap().value;
+        let sum = bins
+            .iter()
+            .map(|sample| sample.value * sample.rate as f64)
+            .sum::<f64>();
+        let count = bins.iter().map(|sample| sample.rate).sum::<u32>();
+        let avg = sum / count as f64;
 
-        let length = samples.len() as f64;
-        let min = *samples.first().unwrap();
-        let max = *samples.last().unwrap();
+        for i in 1..bins.len() {
+            bins[i].rate += bins[i - 1].rate;
+        }
 
-        let median = samples[(0.50 * length - 1.0).round() as usize];
+        let median = find_quantile(&bins, 0.5);
         let quantiles = quantiles
             .iter()
-            .map(|&p| {
-                let sample = samples[(p * length - 1.0).round() as usize];
-                (p, sample)
-            })
+            .map(|&p| (p, find_quantile(&bins, p)))
             .collect();
 
-        let sum = samples.iter().sum();
-        let avg = sum / length;
-
         Some(Self {
             min,
             max,
             median,
             avg,
             sum,
-            count: samples.len() as u64,
+            count: count as u64,
             quantiles,
         })
     }
 }
 
+/// `bins` is a cumulative histogram
+/// We are using R-3 (without choosing the even integer in the case of a tie),
+/// it might be preferable to use a more common function, such as R-7.
+///
+/// List of quantile functions:
+/// https://en.wikipedia.org/wiki/Quantile#Estimating_quantiles_from_a_sample
+fn find_quantile(bins: &[Sample], p: f64) -> f64 {
+    let count = bins.last().expect("bins is not empty").rate;
+    find_sample(bins, (p * count as f64).round() as u32)
```

Comment:
Array indices are usually `usize`, so that is probably preferred here over `u32`.

Reply:
It is used in binary search on `.rate`, which is u32, so I don't think we should use usize here.

Reply:
Got it